rfcs/convergence/simplify-prop-merging.md

```diff
@@ -0,0 +1,359 @@
+
+This proposal aims to tackle the following problems more fundamentally in the design of shorthands and `mergeProps` itself:
+
+- `as` prop is broken when passed to JSX elements through shorthand slots
+- No type safety declaring HTML attributes with `as`
+- Unsafe generic typings for shorthand, that require complex helper types above the internal implementation to make it work
+- Deep merges are costly
+- className merges are not done with `mergeClasses`, which handles scoped directional styles and dedupes atomic declarations
+- Functions are not really merged but simply replaced similar to native object spreading
+
+```tsx
+const props = { button: { customHandleClick /* a custom handler */} };
+const state = mergeProps(
+  { button: { handleClick /* a require handler by our component */ } },
+   resolveShorthandProps(props),
+);
+// Results in => { button: { customHandleClick /* 💣 handlers are not merged */ } }
+```
+
+- When merging different types, simply replaces most recent usage. Although not a problem with slots because of `resolveShorthandProps`
+
+## Problem statement
+
+### Broken `as` prop
+
+Let's take a simple example with a `Loader` component slot inside a `Button` component.
+
+```tsx
+// `Button` state hook
+export const useButton = () => {
+  const state = mergeProps({
+    // ...
+    // Loader component used as shorthand slot
+    loader: { as: Loader },
+  });
+};
+
+// `Button` render function
+export function renderButton(state) {
+  const { slots, slotProps } = getSlots(state);
+  return (
+    <button>
+      {/* ... */}
+      {/* 👇 "slots.loader" will render Loader component */}
+      <slots.loader {...slotProps.loader} />
+    </button>
+  );
+}
+```
+
+```tsx
+// During usage, want to use `Loader` as `div`
+<Button loader={{ as: 'div' }} />
+```
+
+The result is rather worrying:
+
+```tsx
+// Expected result
+<Button>
+  <Loader as="div" />
+</Button>
+
+// Result
+<Button>
+  {/* 👇 this is just a plain div, Loader is not rendered 🤨 */}
+  <div />
+</Button>
+```
+
+In the above example `as` is intended to represent the `Loader` prop, not to define the component to be rendered. This is a result of special handling of the `as` prop which will always be omitted from the slot props and used as the component to be rendered.
+
+### Complex typings and unsafe internal implementation
+
+[microsoft/fluentui#17508](https://github.com/microsoft/fluentui/pull/17508/files) includes type fixes for `mergeProps` which provides extra type safety to validate shorthand slots.
+
+However, the result still leads to complicated component state types benefiting from the stricter typing through generics.
+
+### Native attribute filtering needs to happen by default
+
+For slots that render native elements, it's difficult to provide type safety in order to avoid unsafe DOM attribute spreads as shown in the below example:
+
+```tsx
+// ComponentState
+const state = {
```

Comment:
I'd consider to make this an interface. it might be confusing to readers why is this an actual const.

```
interface ComponentState {
// ...
}
```